server/src/main/java/org/elasticsearch/action/support/FanOutListener.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static <T> boolean tryComplete(Object refValue, ActionListener<T> listener) {
+        if (refValue instanceof SuccessResult result) {
+            result.complete(listener);
+            return true;
+        }
+        if (refValue instanceof FailureResult result) {
+            result.complete(listener);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * If incomplete, atomically update {@link #ref} with the given result and use it to complete any pending listeners.
+     */
+    @SuppressWarnings("unchecked")
+    private void setResult(Object result) {
+        assert isDone(result);
+
+        var currentValue = ref.get();
+        while (true) {
+            if (isDone(currentValue)) {
+                // already complete - nothing to do
+                return;
+            }
+
+            final var witness = ref.compareAndExchange(currentValue, result);
+            if (witness == currentValue) {
+                // we won the race to complete the listener
+                if (currentValue instanceof ActionListener<?> listener) {
+                    // unique subscriber - complete it
+                    var completed = tryComplete(result, listener);
+                    assert completed;
+                } else if (currentValue instanceof Cell currCell) {
+                    // multiple subscribers, but they are currently in reverse order of subscription so reverse them back
+                    Cell prevCell = null;
+                    while (true) {
+                        final var nextCell = currCell.next;
+                        currCell.next = prevCell;
+                        if (nextCell == null) {
+                            break;
+                        }
+                        prevCell = currCell;
+                        currCell = nextCell;
+                    }
+                    // now they are in subscription order, complete them
+                    while (currCell != null) {
+                        // noinspection unchecked
+                        var completed = tryComplete(result, (ActionListener<T>) currCell.listener);
+                        assert completed;
+                        currCell = currCell.next;
+                    }
+                } else {
+                    assert currentValue == EMPTY : "unexpected witness: " + currentValue;
+                }
+                return;
+            }
+
+            // we lost a race with another setResult or addListener call - retry
+            currentValue = witness;
+        }
+    }
+
+    private static boolean isDone(Object refValue) {
+        return refValue instanceof FanOutListener.SuccessResult<?> || refValue instanceof FanOutListener.FailureResult;
+    }
+
+    /**
+     * A cell in the linked list of pending listeners.
+     */
+    private static class Cell {
```

Comment:
Making it accept an `ActionListener<T>` apparently _increases_ the number of casts required, because we create it with `firstListener` captured as the raw type in `currentValue instanceof ActionListener firstListener`, and I couldn't work out how to make it happy with `instanceof ActionListener<T>` here instead. So I'd rather not do that.